File: Hw-1/task7.1/Task7.1/tree.cpp
#include "stdafx.h"
#include "tree.h"
#include <iostream>

using namespace std;

struct Tree
{
	int value;
	Tree *leftChild;
	Tree *rightChild;
};

void insert(Tree *&root, int valueForInsert)
{
	if (root == nullptr)
	{
		Tree *newElement = new Tree{ valueForInsert, nullptr, nullptr};
		root = newElement;
		return;
	}
	if (valueForInsert < root->value)
	{
		insert(root->leftChild, valueForInsert);
	}
	if (valueForInsert > root->value)
	{
		insert(root->rightChild, valueForInsert);
	}
}

bool search(Tree *root, int valueForFind)
{
	if (root == nullptr)
	{
		return false;
	}
	if (valueForFind > root->value)
	{
		return search(root->rightChild, valueForFind);
	}
	if (valueForFind < root->value)
	{
		return search(root->leftChild, valueForFind);
	}
	return true;
}

bool deleteElement(Tree *&root, int valueForDelete)
{
	if (!search(root, valueForDelete))
	{
		return false;
	}
	if (valueForDelete < root->value)
	{
		deleteElement(root->leftChild, valueForDelete);
	}
	if (valueForDelete > root->value)
	{
		deleteElement(root->rightChild, valueForDelete);
	}

	if (valueForDelete == root->value)
	{

		if ((root->leftChild == nullptr) && (root->rightChild == nullptr))
		{
			delete root;
			root == nullptr;
			return true;
		}

		if (root->leftChild == nullptr)
		{
			Tree *oldElement = root;
			root = oldElement->rightChild;
			delete oldElement;
			return true;
		}

		if (root->rightChild == nullptr)
		{
			Tree *oldElement = root;
			root = oldElement->leftChild;
			delete oldElement;
			return true;
		}

		Tree *newroot = root->rightChild;
		while (newroot->leftChild != nullptr)
		{
			newroot = newroot->leftChild;
		}

		root->value = newroot->value;
		newroot->value = valueForDelete;
		deleteElement(root, valueForDelete);
	}
}

void increasePrint(Tree *root)
{
	if (root == nullptr)
	{
		return;
	}
	increasePrint(root->leftChild);
	cout << root->value << " ";
	increasePrint(root->rightChild);
}

void decreasePrint(Tree *root)
{
	if (root == nullptr)
	{
		return;
	}
	decreasePrint(root->rightChild);
	cout << root->value << " ";
	decreasePrint(root->leftChild);
}

void deleteTree(Tree *&root)
{
	if (root == nullptr)
	{
		return;
	}
	if ((root->rightChild == nullptr) && (root->leftChild == nullptr))
	{
		delete root;
		return;
	}
	deleteTree(root->leftChild);
	deleteTree(root->rightChild);
	delete root;
	
	return;
}

File: Hw-2/Homework7/7.2/7.2/Tree.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace _7._2
{
    /// <summary>
    /// класс Дерево
    /// </summary>
    public class Tree<T>
    {
        /// <summary>
        /// класс Элемент дерева
        /// </summary>
        private class TreeElement
        {
            /// <summary>
            /// левый сын
            /// </summary>
            public TreeElement LeftChild { get; set; }

            /// <summary>
            /// правый сын
            /// </summary>
            public TreeElement RightChild { get; set; }

            /// <summary>
            /// значение
            /// </summary>
            public T Value { get; set; }

            /// <summary>
            /// конструктор, создающий элемент дерева
            /// </summary>
            /// <param name="value">значение</param>
            /// <param name="left">левый сын</param>
            /// <param name="right">правый сын</param>
            public TreeElement (T value, TreeElement left, TreeElement right)
            {
                this.Value = value;
                this.LeftChild = left;
                this.RightChild = right;
            }
        }

        /// <summary>
        /// корень дерева
        /// </summary>
        private TreeElement root;

        private void Insert(T value, TreeElement parentLocalRoot, bool right, TreeElement localRoot)
        {
            if ((parentLocalRoot == null) && (localRoot == null))
            {
                TreeElement newElement = new TreeElement(value, null, null);
                this.root = newElement;
                return;
            }
            if (localRoot == null)
            {
                TreeElement newElement = new TreeElement(value, null, null);
                localRoot = newElement;
                if (right)
                {
                    parentLocalRoot.RightChild = localRoot;
                }
                else
                {
                    parentLocalRoot.LeftChild = localRoot;
                }
                return;
            }
            if (String.Compare(value.ToString(), localRoot.Value.ToString()) < 0)
            {
                Insert(value, localRoot, false, localRoot.LeftChild);
            }
            if (String.Compare(value.ToString(), localRoot.Value.ToString()) > 0)
            {
                Insert(value, localRoot, true, localRoot.RightChild);
            }
        }

        /// <summary>
        /// добавить значение в дерево
        /// </summary>
        /// <param name="value">значение</param>
        public bool Add(T value)
        {
            if (IsContaining(value))
            {
                return false;
            }
            this.Insert(value, null, false, root);
            return true;
        }

        private bool Search(TreeElement localRoot, T value)
        {
            if (localRoot == null)
            {
                return false;
            }
            if (String.Compare(value.ToString(), localRoot.Value.ToString()) > 0)
            {
                return Search(localRoot.RightChild, value);
            }
            if (String.Compare(value.ToString(), localRoot.Value.ToString()) < 0)
            {
                return Search(localRoot.LeftChild, value);
            }
            return true;
        }

        /// <summary>
        /// проверить, содержится ли значение в дереве
        /// </summary>
        /// <param name="value">значение</param>
        /// <returns>значение содержится в дереве</returns>
        public bool IsContaining(T value) => Search(root, value);

        private void DelElement(TreeElement parentLocalRoot, bool right, TreeElement localRoot, T valueForDelete)
        {
            if (localRoot == null)
            {
                return;
            }

            if (String.Compare(valueForDelete.ToString(), localRoot.Value.ToString()) < 0)
            {
                DelElement(localRoot, false, localRoot.LeftChild, valueForDelete);
            }
            if (String.Compare(valueForDelete.ToString(), localRoot.Value.ToString()) > 0)
            {
                DelElement(localRoot, true, localRoot.RightChild, valueForDelete);
            }

            if (valueForDelete.Equals(localRoot.Value))
            {

                if ((localRoot.LeftChild == null) && (localRoot.RightChild == null))
                {
                    if (right)
                    {
                        parentLocalRoot.RightChild = null;
                    }
                    else
                    {
                        parentLocalRoot.LeftChild = null;
                    }
                    return;
                }

                if (localRoot.LeftChild == null)
                {
                    TreeElement oldElement = localRoot;
                    localRoot = oldElement.RightChild;
                    if (right)
                    {
                        parentLocalRoot.RightChild = localRoot;
                    }
                    else
                    {
                        parentLocalRoot.LeftChild = localRoot;
                    }
                    return;
                }

                if (localRoot.RightChild == null)
                {
                    TreeElement oldElement = localRoot;
                    localRoot = oldElement.LeftChild;
                    if (right)
                    {
                        parentLocalRoot.RightChild = localRoot;
                    }
                    else
                    {
                        parentLocalRoot.LeftChild = localRoot;
                    }
                    return;
                }

                TreeElement newroot = localRoot.RightChild;
                while (newroot.LeftChild != null)
                {
                    newroot = newroot.LeftChild;
                }

                localRoot.Value = newroot.Value;
                newroot.Value = valueForDelete;
                DelElement(localRoot, false, localRoot, valueForDelete);
            }
        }

        /// <summary>
        /// удалить значение из дерева
        /// </summary>
        /// <param name = "value" > значение </ param >
        /// < returns > значение содержалось в дереве</returns>
        public bool DeleteElement(T value)
        {
            if (!Search(root, value))
            {
                return false;
            }
            DelElement(root, false, root, value);
            return true;
        }

        /// <summary>
        /// проверить, пусто ли дерево
        /// </summary>
        /// <returns>дерево пусто</returns>
        public bool IsEmpty() => root == null;

        private void InOrder(TreeElement localRoot, Stack<T> stack)
        {
            if (localRoot == null)
            {
                return;
            }
            InOrder(localRoot.LeftChild, stack);
            stack.Push(localRoot.Value);
            InOrder(localRoot.RightChild, stack);
        }

        /// <summary>
        /// обход дерева
        /// </summary>
        public IEnumerable<T> Order()
        {
            Stack<T> stack = new Stack<T>();

            InOrder(root, stack);
            int count = stack.Count();
            for (int i = 0; i < count; i++)
            {
                yield return stack.Pop();
            }
        }
    }
}
